component/config.go

```diff
@@ -31,46 +33,131 @@ type ConfigValidator interface {
 // ValidateConfig validates a config, by doing this:
 //   - Call Validate on the config itself if the config implements ConfigValidator.
 func ValidateConfig(cfg Config) error {
-	return validate(reflect.ValueOf(cfg))
+	var err error
+	errs := validate(reflect.ValueOf(cfg))
+
+	for _, suberr := range errs {
+		if suberr.err != nil {
+			if suberr.path != "" {
+				err = errors.Join(err, fmt.Errorf("%s: %w", suberr.path, suberr.err))
+			} else {
+				err = errors.Join(err, suberr.err)
+			}
+		}
+	}
+
+	return err
+}
+
+type pathError struct {
+	err  error
+	path string
 }
 
-func validate(v reflect.Value) error {
+func validate(v reflect.Value) []pathError {
+	errs := []pathError{}
 	// Validate the value itself.
 	switch v.Kind() {
 	case reflect.Invalid:
 		return nil
 	case reflect.Ptr, reflect.Interface:
 		return validate(v.Elem())
 	case reflect.Struct:
-		var errs error
-		errs = multierr.Append(errs, callValidateIfPossible(v))
+		errs = append(errs, pathError{err: callValidateIfPossible(v), path: ""})
 		// Reflect on the pointed data and check each of its fields.
 		for i := 0; i < v.NumField(); i++ {
 			if !v.Type().Field(i).IsExported() {
 				continue
 			}
-			errs = multierr.Append(errs, validate(v.Field(i)))
+			subpathErrs := validate(v.Field(i))
+			for _, err := range subpathErrs {
+				field := v.Type().Field(i)
+				var fieldName string
+				if tag, ok := field.Tag.Lookup("mapstructure"); ok {
+					tags := strings.Split(tag, ",")
+					if len(tags) > 0 {
+						fieldName = tags[0]
+					}
+				}
+				// Even if the mapstructure tag exists, the field name may not
+				// be available, so set it if it is still blank.
+				if len(fieldName) == 0 {
+					fieldName = strings.ToLower(field.Name)
+				}
```

Comment:
Couldn't this be done outside of this loop?

Reply:
Great catch, I've moved it outside the loop.